File: java/src/array/DynamicArray.java
package array;

public class DynamicArray<E> {
    private E[] arr;
    
    private int size;

    private int capacity;

    public String toString() {
	if (this.arr == null)
	    return "";
	
	StringBuffer sb = new StringBuffer();
	for(E data : arr) {
	    if (data != null) {
	        sb.append(data);
                sb.append(",");
	    }
	}

	return sb.toString();
    }


    public int size() {
        return this.size;
    }

    public int capacity() {
	return this.capacity;
    }

    public void pushBack(E data) {
	if (this.size == this.capacity) {
	    this.capacity = this.capacity == 0 ? 1 : this.capacity * 2;
	    E[] newArr = (E[]) new Object[this.capacity];

	    for (int i = 0 ; i < this.size ; i++) {
		newArr[i] = this.arr[i];
	    }
	    this.arr = newArr;
	}
	this.arr[this.size++] = data;
    }

    public E get(int idx) {
	if (idx < 0 || idx >= this.size)
	    throw new IndexOutOfBoundsException();
	return this.arr[idx];
    }

    public void set(int idx, E data) {
	if (idx < 0 || idx >= this.size)
	    throw new IndexOutOfBoundsException();
	this.arr[idx] = data;
    }

    public E remove(int idx) {
	if (idx < 0 || idx >= this.size)
	    return null;
        
	E result = this.arr[idx];

	for (int i = idx ; i < this.size - 1 ; i++)
	    this.arr[i] = this.arr[i + 1];
        this.size--;

	return result;
    }
}

File: python/array/dynamicArray.py
class DynamicArray:
    def __init__(self):
        self.arr = None
        self.size = 0
        self.capacity = 0

    def toString(self):
        if self.arr is None:
            return ''
        else:
            return ','.join([str(data) for data in self.arr if data is not None])

    def length(self):
        return self.size

    def cap(self):
        return self.capacity

    def get(self, idx):
        if idx < 0 or idx >= self.size:
            raise IndexError
        return self.arr[idx]

    def set(self, idx, input):
        if idx < 0 or idx >= self.size:
            raise IndexError
        self.arr[idx] = input

    def pushBack(self, input):
        if self.size == self.capacity:
            length = 1 if self.capacity == 0 else self.capacity * 2
            newArray = [None] * length
            
            for i in range(self.size):
                newArray[i] = self.arr[i]

            self.arr = newArray
            self.capacity = length
        
        self.arr[self.size] = input
        self.size += 1

    def remove(self, idx):
        if idx < 0 or idx >= self.size:
            return None
        
        result = self.arr[idx]
        for i in range(idx, self.size - 1):
            self.arr[i] = self.arr[i + 1]
        self.size -= 1
        return result



